### DISCUSSION TITLE
Cross Dialect Compatibility

### DISCUSSION THREAD

**Discussion by grepusername** [sqlalchemy/sqlalchemy#7199] [Usage Questions] [answered]
Is there a guide anywhere or any recommended approach for how to handle cross-dialect compatibility?

Thinking, in particular, of whether there is a strategy for implementing support for postgres and sqlite on conflict actions, with fallback mechanisms for those backends that do not support such functionality?

At present I am lost at sea with regard to how to elegantly choose when to use the postgres specific insert, it looks like the sqlite on conflict behaviour is set on the column level, etc.  How do I do this like a wizard instead of the sorcerer's apprentice?

EDIT: That is to say, I understand WHEN to use the postgres specific insert, but not how I ought detect it, or at which point in the logic etc. 

Atternatively, it might be the case that this is an ill-advised undertaking and we can either support on-conflict handling or go the universal route?

**Comment by zzzeek** [comment, marked as answer]
so, writing SQLAlchemy code that is cross-dialect compatible for some set of dialects Q is a hard problem, and there is no algorithmic answer to this.    We assume that non-trivial applications are constructed using some level of test-driven-design, or at least write-tests-as-you-go, so that if you for example have a use case that makes use of ON CONFLICT, you'd write a test for it (let's say it's a function), then test it against your target databases, in this case postgresql and SQLite.  You would then address the capabilities or lack thereof for each backend in your function until it works on both platforms.

in this specific case, we have two separate "on conflict" insert() constructs that are used, so your "do_thing_that_uses_on_conflict()" function will have to know up front which backend it's targeting.   When you know this is when you have a database Connection object or a Session to work with, and you look at "connection.engine.name" or "session.connection().engine.name" to see which backend you're working with.   A simplistic way of doing this might look like:


```py
def do_thing_that_uses_on_conflict(connection, data):
    if connection.engine.name == "sqlite":
        # do sqlite logic
    elif connection.engine.name == "postgresql":
        # do postgresql logic
    else:
         raise AssertionError("unsupported backend: %s" % connection.engine.name)
```

I tend to want to automate the above thing using a decorator approach I find useful, a very polished version of it can be seen in oslo db [here](https://github.com/openstack/oslo.db/blob/d7742015f732625ff0c3e455ac5e1c01ba3d9a45/oslo_db/sqlalchemy/utils.py#L880).   

A miniature version of what that oslo.db utility does looks like:

```py

_functions = {}


def dispatch(database_name):
    def go(fn):
        dispatch = _functions.setdefault(fn.__name__, {})
        dispatch[database_name] = fn

        def run(connection, *arg, **kw):
            try:
                real_fn = dispatch[connection.engine.name]
            except KeyError:
                raise KeyError(
                    "unsupported backend for function %s: %s"
                    % (fn.__name__, connection.engine.name)
                )
            else:
                return real_fn(connection, *arg, **kw)

        return run

    return go

```

a proof of concept:

```py
_functions = {}


def dispatch(database_name):
    def go(fn):
        dispatch = _functions.setdefault(fn.__name__, {})
        dispatch[database_name] = fn

        def run(connection, *arg, **kw):
            try:
                real_fn = dispatch[connection.engine.name]
            except KeyError:
                raise KeyError(
                    "unsupported backend for function %s: %s"
                    % (fn.__name__, connection.engine.name)
                )
            else:
                return real_fn(connection, *arg, **kw)

        return run

    return go


from sqlalchemy import create_engine


@dispatch("sqlite")
def run_my_function(connection, arg):
    print("this is a sqlite connection, argument is %s" % arg)


@dispatch("postgresql")
def run_my_function(connection, arg):
    print("this is a postgresql connection, argument is %s" % arg)


db1 = create_engine("sqlite://")
db2 = create_engine("postgresql://scott:tiger@localhost/test")


with db1.connect() as conn:
    run_my_function(conn, "some_Argument")

with db2.connect() as conn:
    run_my_function(conn, "some_Argument")
```

The openstack suite of applications are a very large set of SQLAlchemy apps that target mysql and sqlite simultaneously, with oslo.db being the base SQLAlchemy mediation layer, so I would encourage taking a look at it for ideas.

**Reply by grepusername** [reply]
That's wonderful and very helpful, thank you very much.  It is exactly what I was hoping for.
